### PR TITLE
(chore): benchmark different event processor configurations

### PR DESCRIPTION
**Pull request by thomaszurkan-optimizely**
- default queue size should allow for a app to take a load out of the box. Changed to 1k
- benchmark different queue sizes, queue types, and batch sizes.
- discovered a potential race condition with logger.  that is fixed here is as well.

### CODE REVIEW

File: pkg/event/processor_test.go [hunk 1]
```diff
@@ -449,3 +449,39 @@ func TestChanQueueEventProcessor_ProcessBatch(t *testing.T) {
...
+		WithEventDispatcher(NewMockDispatcher(100, false)))
+	processor.Start(exeCtx)
+
+	impression := BuildTestImpressionEvent()
+	conversion := BuildTestConversionEvent()
+
+	for i := 0; i < b.N; i++ {
+		processor.ProcessEvent(impression)
+		processor.ProcessEvent(conversion)
+		time.Sleep(1 * time.Millisecond)
```
Comment (mikecdavis) [inline pkg/event/processor_test.go]: Doesn't adding the `time.Sleep` nullify the benchmark?

File: pkg/event/processor_test.go [hunk 2]
```diff
@@ -449,3 +449,39 @@ func TestChanQueueEventProcessor_ProcessBatch(t *testing.T) {
 		assert.True(t, len(logEvent.Event.Visitors) >= 1)
 	}
 }
+
+func BenchmarkProcessor(b *testing.B) {
+	exeCtx := utils.NewCancelableExecutionCtx()
+	processor := NewBatchEventProcessor(
+		WithQueueSize(1000),
```
Comment (mikecdavis) [inline pkg/event/processor_test.go]: Following the example in [this](https://dave.cheney.net/2013/06/30/how-to-write-benchmarks-in-go) article you can parameterize the benchmark.
```go
func benchmarkProcessor(qs int, b *testing.B) {
	exeCtx := utils.NewCancelableExecutionCtx()
	processor := NewBatchEventProcessor(
		WithQueueSize(qs)
		WithFlushInterval(10 * time.Millisecond),
		WithEventDispatcher(NewMockDispatcher(100, false)))
	processor.Start(exeCtx)

	impression := BuildTestImpressionEvent()
	conversion := BuildTestConversionEvent()

	for i := 0; i < b.N; i++ {
		processor.ProcessEvent(impression)
		processor.ProcessEvent(conversion)
		time.Sleep(500)
	}
	processor.Stop()
}

func BenchmarkQueueSize100(b *testing.B) { benchmarkProcessor(100, b) }
func BenchmarkQueueSize2000(b *testing.B) { benchmarkProcessor(2000, b) }
func BenchmarkQueueSize5000(b *testing.B) { benchmarkProcessor(5000, b) }
func BenchmarkQueueSize10000(b *testing.B) { benchmarkProcessor(10000, b) }
func BenchmarkQueueSize50000(b *testing.B) { benchmarkProcessor(50000, b) }
```

File: pkg/event/processor_test.go [hunk 3]
```diff
@@ -449,3 +449,39 @@ func TestChanQueueEventProcessor_ProcessBatch(t *testing.T) {
...
+		time.Sleep(1 * time.Millisecond)
+	}
+}
+
+func BenchmarkProcessorDefault(b *testing.B) {
+	exeCtx := utils.NewCancelableExecutionCtx()
+	processor := NewBatchEventProcessor(
+		WithFlushInterval(10 * time.Millisecond),
+		WithEventDispatcher(NewMockDispatcher(100, false)))
+	processor.Start(exeCtx)
```
Comment (mikecdavis) [inline pkg/event/processor_test.go]: Assuming we should `Stop()` as well?

File: pkg/event/processor_test.go (hunk 1 above)
Comment (thomaszurkan-optimizely) [reply pkg/event/processor_test.go]: There is literally no other way to let the go routines run that put in a sleep.  The sleep is now 500 nanoseconds.

File: pkg/event/processor_test.go [hunk 4]
```diff
@@ -449,3 +448,153 @@ func TestChanQueueEventProcessor_ProcessBatch(t *testing.T) {
...
+func BenchmarkWithBatchSize(b *testing.B) {
+	logging.SetLogger(&NoOpLogger{})
+
+	merges := []struct {
+		name string
+		batchSize int
+		fun  func(bs int, b *testing.B) int
+	}{
+		{"BatchSize10", 10, benchmarkProcessorWithBatchSize},
+		{"BatchSize20", 20,benchmarkProcessorWithBatchSize},
```
Comment (mikecdavis) [inline pkg/event/processor_test.go]: Nit, spacing

File: pkg/event/processor_test.go [hunk 5]
```diff
@@ -449,3 +448,153 @@ func TestChanQueueEventProcessor_ProcessBatch(t *testing.T) {
...
+	benchmarkProcessorQueueSize(2000, b)
+}
+
+func benchmarkProcessor3000(b *testing.B) {
+	benchmarkProcessorQueueSize(3000, b)
+}
+
+func benchmarkProcessor4000(b *testing.B) {
+	benchmarkProcessorQueueSize(4000, b)
+}
```
Comment (mikecdavis) [inline pkg/event/processor_test.go]: Given how you are parameterizing within `BenchmarkWithQueueSize` you can get rid of these functions.

File: pkg/event/processor_test.go [hunk 6]
```diff
@@ -449,3 +448,153 @@ func TestChanQueueEventProcessor_ProcessBatch(t *testing.T) {
...
+
+	conversion := BuildTestConversionEvent()
+
+	for i := 0; i < b.N; i++ {
+		processor.ProcessEvent(conversion)
+	}
+
+	exeCtx.TerminateAndWait()
+
+	return dispatcher.Events.Size()
```
Comment (mikecdavis) [inline pkg/event/processor_test.go]: Are these `return` values use for anything?

File: pkg/logging/logger.go [hunk 7]
```diff
@@ -46,17 +48,23 @@ const (
 )
 
 func init() {
+	mutex.Lock()
 	defaultLogConsumer = NewFilteredLevelLogConsumer(LogLevelInfo, os.Stdout)
+	mutex.Unlock()
```
Comment (mikecdavis) [inline pkg/logging/logger.go]: I know this was already there, but is this `init()` needed or is it preferable to set during definition?
```go
var defaultLogConsumer OptimizelyLogConsumer = NewFilteredLevelLogConsumer(LogLevelInfo, os.Stdout)
```

File: pkg/logging/logger.go [hunk 8]
```diff
@@ -46,17 +48,23 @@ const (
...
 }
 
 // SetLogger replaces the default logger with the given logger
 func SetLogger(logger OptimizelyLogConsumer) {
-	defaultLogConsumer = logger
+	mutex.Lock()
+	 defaultLogConsumer = logger
```
Comment (mikecdavis) [inline pkg/logging/logger.go]: Nit, spacing.

File: pkg/logging/logger.go [hunk 9]
```diff
@@ -30,6 +31,7 @@ func (l LogLevel) String() string {
 }
 
 var defaultLogConsumer OptimizelyLogConsumer
+var mutex = &sync.Mutex{}
```
Comment (mikecdavis) [inline pkg/logging/logger.go]: Curious what issue was seen that we needed to protect the logger. The operations below seem to be atomic and shouldn't require a mutex, but I might be missing something.

File: pkg/event/processor_test.go [hunk 10]
```diff
@@ -449,3 +448,153 @@ func TestChanQueueEventProcessor_ProcessBatch(t *testing.T) {
...
+func (l *NoOpLogger) Log(level logging.LogLevel, message string, fields map[string]interface{}) {
+}
+
+
+func (l *NoOpLogger) SetLogLevel(level logging.LogLevel) {
+
+}
+
+func BenchmarkWithQueueSize(b *testing.B) {
+	logging.SetLogger(&NoOpLogger{})
```
Comment (mikecdavis) [inline pkg/event/processor_test.go]: Can you elaborate on why we needed to set a `NoOpLogger`?

File: pkg/logging/logger.go (hunk 7 above)
Comment (thomaszurkan-optimizely) [reply pkg/logging/logger.go]: since there is a SetLogger.  I actually think this is safer.

File: pkg/logging/logger.go (hunk 9 above)
Comment (thomaszurkan-optimizely) [reply pkg/logging/logger.go]: so, setLogger can be called at anytime.  During benchmarking, I was calling set at the begging and resetting at the end.  This caused a race condition which is now resolved.

File: pkg/event/processor_test.go (hunk 6 above)
Comment (thomaszurkan-optimizely) [reply pkg/event/processor_test.go]: yes, they are now used in the queue size tests to make sure that all events are processed.

File: pkg/event/processor.go [hunk 11]
```diff
@@ -33,7 +33,7 @@ import (
 
 // Processor processes events
 type Processor interface {
-	ProcessEvent(event UserEvent)
+	ProcessEvent(event UserEvent) bool
```
Comment (mikeng13) [inline pkg/event/processor.go]: Is the bool return value ever used? I don’t see why this is necessary unless the client makes use of it

File: pkg/event/processor_test.go (hunk 3 above)
Comment (thomaszurkan-optimizely) [reply pkg/event/processor_test.go]: Added stop via context terminate.

File: pkg/event/processor.go (hunk 11 above)
Comment (mikeng13) [reply pkg/event/processor.go]: 👍

File: go.mod [hunk 12]
```diff
@@ -3,13 +3,17 @@ module github.com/optimizely/go-sdk
 go 1.12
 
 require (
+	github.com/DATA-DOG/godog v0.7.13
```
Comment (mikeng13) [inline go.mod]: Can you remove these please? They are used only for FSC testing, which shoudn't be included in our main deps

**Pull request merged**